### ISSUE TITLE
Refactor subs() of multivariate polynomials for readability and efficiency

### ISSUE THREAD

**Issue by kwankyu** [sagemath/sage#34591]
<div id="comment:0"></div>

This bug

```
sage: R = Integers(8)
sage: RXY.<X,Y> = R[]
sage: F = X^3*Y
sage: F.subs({X:1})
Y
sage: F.subs({X:2})  # should be 0!
Y
```
is fixed by #34417. However while fixing the bug, it was noticed that the code of `.subs()` is so convoluted that the code is hard to read and inefficient.

So in this ticket, we refactor the code. The modified code is shorter, simpler, and seems slightly faster. 

Moreover the patch also fixes the original bug without #34417 (upstream patch)!

Before the patch:

```sage
sage: R = Integers(8)
sage: RXY.<X,Y> = R[]
sage: F = X^3*Y
sage: F.subs({X:2})
Y
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.41 ms per loop
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.39 ms per loop
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.41 ms per loop
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.46 ms per loop
sage: timeit('F.subs({X:2})')
125 loops, best of 3: 3.38 ms per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.26 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.36 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.65 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.21 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.09 μs per loop
```
After the patch 

```sage
sage: R = Integers(8)
sage: RXY.<X,Y> = R[]
sage: F = X^3*Y
sage: F.subs({X:2})
0
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.11 μs per loop
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.8 μs per loop
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.63 μs per loop
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.52 μs per loop
sage: timeit('F.subs({X:2})')
625 loops, best of 3: 8.63 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.18 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 7.99 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.25 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.2 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.6 μs per loop
sage: timeit('F.subs({X:1})')
625 loops, best of 3: 8.38 μs per loop
```

CC:  @videlec @yyyyx4

Component: **basic arithmetic**

Author: **Kwankyu Lee**

Branch/Commit: **[u/klee/multivariate_polynomial_variable_substition_has_a_bug](https://github.com/sagemath/sagetrac-mirror/tree/u/klee/multivariate_polynomial_variable_substition_has_a_bug) @ [`4a47cb2`](https://github.com/sagemath/sagetrac-mirror/commit/4a47cb2737817e04f55c478494e3645e8a78272c)**

_Issue created by migration from https://trac.sagemath.org/ticket/34591_

**Comment by kwankyu** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1 +1,10 @@
 
+```
+sage: R = Integers(8)
+sage: RXY.<X,Y> = R[]
+sage: F = X^3*Y
+sage: F.subs({X:1})
+Y
+sage: F.subs({X:2})  # should be 0!
+Y
+```
``````

**Comment by kwankyu** [comment]
<div id="comment:3" align="right">comment:3</div>

This bug is related with these lines

```diff
diff --git a/src/sage/rings/polynomial/multi_polynomial_libsingular.pyx b/src/sage/rings/polynomial/multi_polynomial_libsingular.pyx
index 2109c516a1..35a011174a 100644
--- a/src/sage/rings/polynomial/multi_polynomial_libsingular.pyx
+++ b/src/sage/rings/polynomial/multi_polynomial_libsingular.pyx
@@ -3556,7 +3556,7 @@ cdef class MPolynomial_libsingular(MPolynomial):
 
                 if _p == NULL:
                     # polynomial becomes 0 after some substitution
-                    try_symbolic = 1
+                    try_symbolic = 1  # Why?
                     break
 
         cdef dict gd
```

At line 3557, the substituted polynomial becomes 0 correctly. But then it goes on "try_symbolic". Why?

These lines were introduced in #17785.

**Comment by kwankyu** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -8,3 +8,5 @@
 sage: F.subs({X:2})  # should be 0!
 Y
 ```
+
+Reported in https://groups.google.com/g/sage-support/c/RzHODsb77KU
``````

**Comment by DaveWitteMorris** [comment]
<div id="comment:6" align="right">comment:6</div>

I think this is a duplicate of #34417.

**Comment by kwankyu** [comment]
<div id="comment:7" align="right">comment:7</div>

Replying to [Dave Morris](#comment%3A6):
> I think this is a duplicate of #34417.

Yes. But as I analyzed the bug, there is no problem with Singular. So I suggest to close #34417 as a duplicate, and let us focus on fixing the problem in [comment:3](#comment%3A3).

**Comment by DaveWitteMorris** [comment]
<div id="comment:8" align="right">comment:8</div>

I don't know why you say there is no problem with singular. As pointed out in [#34417 comment:1](https://github.com/sagemath/sage/issues/34417#comment:1), the problem seems to arise in `singular_polynomial_call`.  

I don't think the line in [comment:3](#comment%3A3) is actually the bug (but I may be wrong). My understanding is that it is just setting a flag to say that something has gone wrong, so singular should be used to do the substitution. But it seems that singular gives the wrong answer (or the interface makes a mistake). It may be possible to solve this particular bug report by avoiding the call to singular, but it does seem to have uncovered a problem with singular (or its interface).

**Comment by kwankyu** [comment]
<div id="comment:10" align="right">comment:10</div>

Replying to [Dave Morris](#comment%3A8):
> I don't know why you say there is no problem with singular. As pointed out in [#34417 comment:1](https://github.com/sagemath/sage/issues/34417#comment:1), the problem seems to arise in `singular_polynomial_call`.  

You are right.

But there still seems to be a different issue with `.subs()`. I am investigating.

**Comment by kwankyu** [comment]
Branch: **[u/klee/multivariate_polynomial_variable_substition_has_a_bug](https://github.com/sagemath/sagetrac-mirror/tree/u/klee/multivariate_polynomial_variable_substition_has_a_bug)**

**Comment by kwankyu** [comment]
Commit: **[`4a47cb2`](https://github.com/sagemath/sagetrac-mirror/commit/4a47cb2737817e04f55c478494e3645e8a78272c)**

**Comment by kwankyu** [comment]
<div id="comment:12"></div>

New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/4a47cb2737817e04f55c478494e3645e8a78272c"><code>4a47cb2</code></a></td><td><code>Refactor .subs() for simplicity and efficiency</code></td></tr></table>

**Comment by kwankyu** [comment]
Author: **Kwankyu Lee**

**Comment by kwankyu** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1,3 +1,4 @@
+This bug
 
 ```
 sage: R = Integers(8)
@@ -8,5 +9,71 @@
 sage: F.subs({X:2})  # should be 0!
 Y
 ```
+is fixed by #34417. However while fixing the bug, it was noticed that the code of `.subs()` is so convoluted that the code is hard to read and inefficient.
 
-Reported in https://groups.google.com/g/sage-support/c/RzHODsb77KU
+So in this ticket, we refactor the code. The modified code is shorter, simpler, and seems slightly faster. 
+
+Moreover the patch also fixes the original bug without #34417 (upstream patch)!
+
+Before the patch:
+
+```sage
+sage: R = Integers(8)
+sage: RXY.<X,Y> = R[]
+sage: F = X^3*Y
+sage: F.subs({X:2})
+Y
+sage: timeit('F.subs({X:2})')
+125 loops, best of 3: 3.41 ms per loop
+sage: timeit('F.subs({X:2})')
+125 loops, best of 3: 3.39 ms per loop
+sage: timeit('F.subs({X:2})')
+125 loops, best of 3: 3.41 ms per loop
+sage: timeit('F.subs({X:2})')
+125 loops, best of 3: 3.46 ms per loop
+sage: timeit('F.subs({X:2})')
+125 loops, best of 3: 3.38 ms per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.26 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.36 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.65 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.21 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.09 μs per loop
+```
+After the patch 
+
+```sage
+sage: R = Integers(8)
+sage: RXY.<X,Y> = R[]
+sage: F = X^3*Y
+sage: F.subs({X:2})
+0
+sage: timeit('F.subs({X:2})')
+625 loops, best of 3: 8.11 μs per loop
+sage: timeit('F.subs({X:2})')
+625 loops, best of 3: 8.8 μs per loop
+sage: timeit('F.subs({X:2})')
+625 loops, best of 3: 8.63 μs per loop
+sage: timeit('F.subs({X:2})')
+625 loops, best of 3: 8.52 μs per loop
+sage: timeit('F.subs({X:2})')
+625 loops, best of 3: 8.63 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.18 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 7.99 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.25 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.2 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.6 μs per loop
+sage: timeit('F.subs({X:1})')
+625 loops, best of 3: 8.38 μs per loop
+```
+
+
``````
